backend/uclapi/oauth/views.py

```diff
@@ -0,0 +1,347 @@
+        given_name = request.META['HTTP_GIVENNAME']
+        display_name = request.META['HTTP_DISPLAYNAME']
+        employee_id = request.META['HTTP_EMPLOYEEID']
+    except:
+        context = {
+            "error": "Didn't receive all required Shibboleth data."
+        }
+        return render(
+            request,
+            'shibboleth_error.html',
+            context=context,
+            status=400
+        )
+
+    # If a user has never used the API before then we need to sign them up
+    try:
+        user = User.objects.get(email=eppn)
+    except ObjectDoesNotExist:
+        # create a new user
+        user = User(
+            email=eppn,
+            full_name=display_name,
+            given_name=given_name,
+            department=department,
+            cn=cn,
+            raw_intranet_groups=groups,
+            employee_id=employee_id
+        )
+
+        user.save()
+        keen.add_event("signup", {
+            "id": user.id,
+            "email": eppn,
+            "name": display_name
+        })
+    
+    signer = TimestampSigner()
+    response_data = {
+        "client_id": app.client_id,
+        "state": state,
+        "user_upi": user.employee_id
+    }
+
+    response_data_str = json.dumps(response_data, cls=DjangoJSONEncoder)
+    response_data_signed = signer.sign(response_data_str)
+
+    page_data = {
+        "app_name": app.name,
+        "creator": app.user.full_name,
+        "client_id": app.client_id,
+        "state": state,
+        "scope": {
+            "private_roombookings": app.scope.private_roombookings,
+            "private_timetable": app.scope.private_timetable,
+            "private_uclu": app.scope.private_uclu
+        },
+        "user": {
+            "full_name": user.full_name,
+            "cn": user.cn,
+            "email": user.email,
+            "department": user.department,
+            "upi": user.employee_id
+        },
+        "signed_data": response_data_signed
+    }
+
+    initial_data = json.dumps(page_data, cls=DjangoJSONEncoder)
+    return render(request, 'permissions.html',
+    {
+        'initial_data': initial_data
+    })
+
+
+@csrf_protect
+def userdeny(request):
+    signer = TimestampSigner()
+
+    try:
+        signed_data = request.POST.get("signed_app_data")
+        raw_data_str = signer.unsign(signed_data, max_age=300)
+    except:
+        return JsonResponse({
+            "ok": False,
+            "error": "The signed data received was invalid. Please try the login process again. If this issue persists, please contact support."
+        })
+    
+    try:
+        data = json.loads(raw_data_str)
+    except:
+        return JsonResponse({
+            "ok": False,
+            "error": "The JSON data was not in the expected format. Please contact support."
+        })
+
+    app = App.objects.get(client_id=data["client_id"])
+    state = data["state"]
+
+    redir = app.callback_url + "/denied?state=" + state
```

Comment:
Slack does something like this, but what I mean here is a "we should have one callback, not 4" thing